cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -410,7 +410,7 @@ public void removeApplication(final Id.Application appId) throws Exception {
                        new com.google.common.base.Predicate<ProgramRuntimeService.RuntimeInfo>() {
       @Override
       public boolean apply(ProgramRuntimeService.RuntimeInfo runtimeInfo) {
-        return runtimeInfo.getProgramId().toEntityId().getApplication().equals(appId.getId());
+        return runtimeInfo.getProgramId().toEntityId().getParent().equals(appId.toEntityId());
```

Comment:
Better to use `ApplicationId` in `ProgramRuntimeService.RuntimeInfo` so that the same program running in different versions of the app can be distinguished? I can open another PR for this.